### PR TITLE
Victor

### PR DESCRIPTION
**Pull request by PetervDooren**

### CODE REVIEW

File: ed_sensor_integration/tools/segmenter.cpp [hunk 1]
```diff
@@ -264,79 +273,260 @@ int main(int argc, char **argv)
...
+        EntityRepresentation2D entity_2d = fitter.GetOrCreateEntity2D(e);
+
+        geo::Pose3D pose = e->pose();
+        //float yaw_ent = atan2(pose.R.yx, pose.R.xx)
+
+        // entity yaw
+        float yaw_ent = atan2(pose.R.yx, pose.R.xx);
+
+        // robot yaw
+        float yaw_robot = atan2(snapshot.sensor_pose.R.yx, snapshot.sensor_pose.R.xx);
```
Comment (PetervDooren) [inline ed_sensor_integration/tools/segmenter.cpp]: The reason this simple formula works for the yaw of the entity is that there is only rotation around the gravity axis. The sensor pose is full rotation (x,y,z). Getting its yaw pose is slightly more complex

File: ed_sensor_integration/tools/segmenter.cpp [hunk 2]
```diff
@@ -264,79 +273,260 @@ int main(int argc, char **argv)
...
+                float diff_x1 = x_wm1 - snapshot.sensor_pose.t.x;
+                float diff_y1 = y_wm1 - snapshot.sensor_pose.t.y;
+                float diff_x2 = x_wm2 - snapshot.sensor_pose.t.x;
+                float diff_y2 = y_wm2 - snapshot.sensor_pose.t.y;
+
+                // rotating coordinate system to that of the robot
+                float x_m1 = cos(yaw_robot) * diff_x1 + sin(yaw_robot) * diff_y1;
+                float y_m1 = -sin(yaw_robot) * diff_x1 + cos(yaw_robot) * diff_y1;
+                float x_m2 = cos(yaw_robot) * diff_x2 + sin(yaw_robot) * diff_y2;
+                float y_m2 = -sin(yaw_robot) * diff_x2 + cos(yaw_robot) * diff_y2;
```
Comment (PetervDooren) [inline ed_sensor_integration/tools/segmenter.cpp]: advice, calculate the relative pose of the entity w.r.t. the robot and use that to draw your pictures. Saves a lot of bookkeeping.

**Comment by MatthijsBurgh** [comment]
I don't think we want to merge this. At least not any time soon.

**Comment by PetervDooren** [comment]
correct. this is mostly to give feedback to Victor while allowing me to reference specific lines

File: ed_sensor_integration/tools/segmenter.cpp [hunk 3]
```diff
@@ -184,6 +184,58 @@ void showSegmentationResults(const Snapshot& snapshot, const UpdateRequest& upda
...
+    quaternion.x = geo_quaternion.getX();
+    quaternion.y = geo_quaternion.getY();
+    quaternion.z = geo_quaternion.getZ();
+    quaternion.w = geo_quaternion.getW();
+
+    // Get the Euler angles
+    EulerAngles angles = ToEulerAngles(quaternion);
+    ROS_DEBUG_STREAM("Matrix: " << rotation << " --> yaw: " << angles.yaw);
+    return angles.yaw;
+}
```
Comment (MatthijsBurgh) [inline ed_sensor_integration/tools/segmenter.cpp]: If this is correct code, move it to geolib2

Comment (PetervDooren) [reply ed_sensor_integration/tools/segmenter.cpp]: In the scope of your BEP this is not required but if you would like to learn how to do this let us know. I can also imagine that there may be other libraries that can do this.

**Pull request closed without merging**